rust/src/http2/detect.rs

```diff
@@ -0,0 +1,399 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::http2::{HTTP2FrameTypeData, HTTP2Transaction};
+use super::parser;
+use crate::core::{STREAM_TOCLIENT, STREAM_TOSERVER};
+use std::ffi::CStr;
+use std::mem::transmute;
+use std::str::FromStr;
+
+#[no_mangle]
+pub extern "C" fn rs_http2_tx_get_frametype(
+    tx: *mut std::os::raw::c_void, direction: u8,
+) -> std::os::raw::c_int {
+    let tx = cast_pointer!(tx, HTTP2Transaction);
+    //TODO6tx check use of direction
+    match direction {
+        STREAM_TOSERVER => match &tx.ftype {
+            None => {
+                return -1;
+            }
+            Some(x) => {
+                return *x as i32;
+            }
+        },
+        STREAM_TOCLIENT => match &tx.ftype {
+            None => {
+                return -1;
+            }
+            Some(x) => {
+                return *x as i32;
+            }
+        },
+        _ => {}
+    }
+
+    return -1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_http2_parse_frametype(
+    str: *const std::os::raw::c_char,
+) -> std::os::raw::c_int {
+    let ft_name: &CStr = unsafe { CStr::from_ptr(str) };
```

Comment:
Unless there is a specific reason not to (a define function type or something), I'd just go ahead and mark this function unsafe to avoid the clippy warning.

Reply:
What do you mean ?
What is the clippy warning ?

In general, I try to restrict `unsafe` to the smallest part of the code

Reply:
> In general, I try to restrict `unsafe` to the smallest part of the code

Some discussion on this here: https://github.com/OISF/suricata/pull/4666

Consensus on the Rust side is that if the function uses the raw pointers passed in, the whole thing should be unsafe, and clippy (the Rust linter) will pick this up.

Reply:
Thanks for the interesting pointer.

I will make the change, and make better reviews next time :-)

Is there an easy command to run clippy on my code ?

Reply:
`unsafe` added in current v12